docs/developer/architecture/ci.rst

```diff
@@ -0,0 +1,122 @@
+CI
+==
+
+Pre-merge
+---------
+
+What should be tested in pre-merge on every ``user/*``, ``feature/*``,
+``improvement/*``, ``bugfix/*``, ``w/*`` branches. The pre-merge test should
+not long too much time (less than 40 minutes is great) so we can't
+test everything in pre-merge, we should only test building of the product
+and check that product still usable.
+
+- Building tests
+
+  - Build
+  - Lint
+  - Unit tests
+
+- Installation tests
+
+  - Simple install RHEL
+  - Simple install CentOs + expansion
+
+When merging several pull requests at the same time, given that we are on a
+queue branch (``q/*``), we may want to have or we may require additional tests
+as combinaison of several PRs could have larger impact than all individual PR.
+
+- Simple upgrade/downgrade
+
+.. _post-merge:
+
+Post-merge
+----------
+
+On each and every ``development/2.*`` branches we want to run complex tests
+to ensure that the current product continues to work.
+
+Nightly
+^^^^^^^
+
+- Upgrade, downgrade tests:
+
+  - For development branch just before
+
+    e.g.: on ``development/2.x`` test upgrade from ``development/2.(x-1)``
+    and downgrade to ``development/2.(x-1)``
+
+    - Build branch ``development/2.(x-1)``
+    - Tests:
+
+      - Single node test
+      - Complex deployment test
+
+  - For last released version of current minor
+
+    e.g.: on ``development/2.x`` when developing "2.x.y" test upgrade from
+    ``metalk8s-2.x.(y-1)`` and downgrade to ``metalk8s-2.x.(y-1)``
+
+    - Single node test
+    - Complex deployment test
+
+  - For last released version of minor just before
+
+    e.g.: on ``development/2.x`` when developing "2.x.y" test upgrade from
+    ``metalk8s-2.(x-1).z`` and downgrade to ``metalk8s-2.(x-1).z`` where
+    "2.(x-1).z" is the last patch released version for "2.(x-1)" (``z`` may be
+    different from ``y``)
+
+    - Single node test
+    - Complex deployment test
+
+- Backup, restore tests
+- Solutions tests
+
+.. todo::
+
+   - Define what is "complex deployment"
+
+Weekly
+^^^^^^
+
+More complex tests (TBD)
+
+Adaptive test plan
+------------------
+
+CI pre-merge may be more flexible by including some logic about the content
+of the changeset.
```

Comment:
Let's remind what the problem is, what the goals are, and why we think adapting the pre-merge test plan is acceptable (queue builds cover everything, post-merge will always run anyway...), before jumping into the implementation (which may not be valid once we compare it with goals/requirements).

Reply:
ping
If you don't want to adress this atm, create a follow-up issue with this comment linked, please

Reply:
Just added a sentence about the goal.

But I don't see what you want to say about "why is acceptable ?" your commit change doc you only want to test doc it's just ... obvious IMO I don't see.
I can add a sentence about the fact that we test the product in post-merge  but btw queue build is not acceptable here because they only run for multi PRs merges (or not uptodate ones)

Reply:
> but btw queue build is not acceptable here because they only run for multi PRs merges (or not uptodate ones)

Yes, but I feel that's a problem. I'd really want anything merged to have been tested with the full pre-merge suite at least once. IMO, adjusting the test plan should be a tool for developers to focus the feedback CI gives them on a given patchset. While the review process is ongoing, focusing tests on what the changes affect allows for quicker iterations. Once approval is given, assuming no flakiness at all in pre-merge tests, adding a full suite in a merge queue shouldn't cause trouble, and safeguard us against merging broken code (of course, that means broken queues, but that's why we put them in place).

Reply:
That mean lot of change in current eve/bert-e work, because here that only check that a commit get success build so you need to make a distinction between build from pre-merge and build for queue both have a success final status for the same commit but one allow to merge and the other one not ....